Add DebugMenuInterface and TestDebugMenu(WIP)

RevySR opened the pull request.

RevySR committed:
Add DebugMenuInterface and TestDebugMenu(WIP)

RevySR committed:
Add TestMenuInterface (broken)

RevySR committed:
Merge remote-tracking branch 'upstream/master' into AddDebugMenuInterface

RevySR committed:
Fix Group count and Finish DebugMenu

RevySR committed:
Clean up uesless code

Seng-Jik commented on Core/DebugMenu.h:
```diff
@@ -0,0 +1,148 @@
+
+			return true;
+		}
+
```
private

Seng-Jik replied:
font和space缩小，作为constexpr，并private
字体的大小和间距参考DebugDisplay

Seng-Jik replied:
默认选择第一项菜单项

Seng-Jik commented on Core/DebugMenu.h:
```diff
@@ -0,0 +1,148 @@
+		typename TEngine,
+		typename TWindow,
+		typename TInput>
+	class DebugMenuInterface: public Object, private NoCopyMove
```
[[nodiscard]]

Seng-Jik committed:
Merge branch 'master' into AddDebugMenuInterface

Seng-Jik commented on Core/DebugMenu.h:
```diff
@@ -0,0 +1,148 @@
+		typename TInput>
+	class DebugMenuInterface: public Object, private NoCopyMove
+	{
+		class[[nodiscard]] TestMenuItemInterface : public UI::TextMenuItemInterface<TFontRenderer>
```
改为DebugMenuItem？

Seng-Jik commented on TestWindows/ObjectTest.cpp:
```diff
@@ -427,3 +427,37 @@ TEST(SceneTest, TestDebugDisplay)
+		Snowing::Log("debugMenu write log2");
+	});
+	
+	g.Emplace<Scene::VirtualTask>(5.0f, [] {Engine::Get().Exit(); });
```
测试时间改为1秒吧。

Seng-Jik commented on Core/DebugMenu.h:
```diff
@@ -0,0 +1,148 @@
+			Enter.Update();
+			if (Enter.JustPress())
+			{
+				menu_.GetSelectedObject().value()->OK();
```
此处应该判空

Seng-Jik commented on Core/DebugMenu.h:
```diff
@@ -0,0 +1,148 @@
+		class[[nodiscard]] TestMenuItemInterface : public UI::TextMenuItemInterface<TFontRenderer>
+		{
+		public:
+			TestMenuItemInterface(
```
我觉得fontSize和space可以写死

RevySR committed:
Modify for review

https://github.com/Strrationalism/Snowing/pull/39/files/32278d4ca61ceb455c7fb7f68e025f601f9ab049

RevySR committed:
Merge branch 'AddDebugMenuInterface' of https://github.com/Rabenda/Snowing into AddDebugMenuInterface

RevySR committed:
Add condition judgement

RevySR committed:
Remove Group::IsEmpty and add CountWithNewObject, revert Count function
